Properly handle H2 upgrade with data packets

isaacrivriv opened the pull request:
Added logic to handle misparsed data on HTTP2 upgrade
- Detect if data is missing when Upgrade comes
- Queue up data locally before handling upgrade
- Reset everything after

isaacrivriv committed:
Added logic to handle missing data on HTTP2 upgrade
- Detect if data is missing when Upgrade comes
- Queue up data locally before handling upgrade
- Reset everything after

isaacrivriv commented:
#build

pmd1nh commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/inbound/HttpInputStreamImpl.java:
```diff
@@ -468,6 +481,27 @@ public void restart() {
         }
     }
 
+    public void resetReaderChannelMultiRead() {
```
May be change the method name to "setReadFromChannelComplete" ?

pmd1nh commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:
```diff
@@ -1338,6 +1339,40 @@ public boolean handleHTTP2UpgradeRequest(Map<String, String> http2Settings) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Got exception reading request and queueing up data. Can't handle request upgrade to HTTP2.", e);
+                    }
+                    vc.getStateMap().put(h2InitError, true);
```
if there is an exception, how is the channel multi read buffer postDataBuffer get cleanup?  It depends on the dataAlreadyReadFromChannel (true) (in the non custom property case) which now can't be set due to an exception.

isaacrivriv committed:
Corrected missing logic for fix
- H2 calls re-init on service context which sets state to headers parsed
- Added fix to set body read after H2 process finishes to correct that
  logic

isaacrivriv commented:
#build

isaacrivriv commented:
#libby

pmd1nh commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java:
```diff
@@ -310,15 +310,15 @@ final public void resetMsgParsedState() {
      * @return boolean
      */
-    final protected boolean isBodyComplete() {
+    final public boolean isBodyComplete() {
```
Update the Copy right year for this class.

pmd1nh commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/inbound/HttpInputStreamImpl.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2009 IBM Corporation and others.
+ * Copyright (c) 2009, 2023 IBM Corporation and others.
```
Copyright 2009, 2024

pmd1nh commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:
```diff
@@ -35,6 +35,7 @@
 import com.ibm.ws.http.channel.internal.inbound.HttpInboundChannel;
 import com.ibm.ws.http.channel.internal.inbound.HttpInboundLink;
 import com.ibm.ws.http.channel.internal.inbound.HttpInboundServiceContextImpl;
+import com.ibm.ws.http.channel.internal.inbound.HttpInputStreamImpl;
```
update Copyright

isaacrivriv committed:
Worked through review comments

isaacrivriv commented:
Build hasn't started so will work with the same build link

isaacrivriv commented:
#libby

isaacrivriv commented:
#build

isaacrivriv commented:
#libby
